File: Story2/Quest1/a.rs
use std::fs;
use std::collections::HashSet;

fn main() {
  let input = fs::read_to_string("input.txt").unwrap();

  let (nails, coins) = process(&input);
  let width = nails.iter().map(|&p| p.x).max().unwrap();
  let height = nails.iter().map(|&p| p.y).max().unwrap();

  let mut total = 0;
  for (x, coin) in (1..=9).map(|x| (x - 1) * 2).zip(coins) {
    let mut p = Point { x, y: 0 };
    let mut c = coin.chars().cycle();

    while p.y <= height {
      if nails.contains(&p) {
        let d = c.next().unwrap();
        p.x += match p.x {
          0 => 1,
          _ if p.x == width => -1,
          _ => if d == 'R' { 1 } else { -1 },
        }
      }

      p.y += 1;
    }

    let score = (p.x / 2 + 1) * 2 - (x / 2 + 1);
    total += if score > 0 { score } else { 0 };
  }

  println!("{total}");
}

fn process(input: &str) -> (HashSet<Point>, Vec<String>) {
  let mut blocks = input.split("\n\n");

  let grid = blocks.next().unwrap();
  let mut nails: HashSet<Point> = HashSet::new();
  for (i, line) in grid.lines().enumerate() {
    for (j, c) in line.chars().enumerate() {
      if c == '*' { nails.insert(Point { x: j as i32, y: i as i32 }); }
    }
  }

  let coins: Vec<String> = blocks.next().unwrap().lines().map(|s| String::from(s)).collect();

  (nails, coins)
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Point {
  x: i32,
  y: i32,
}

File: Story2/Quest1/a.swift
struct Point: Equatable, Hashable {
  var x: Int
  var y: Int

  init(x: Int, y: Int) {
    self.x = x
    self.y = y
  }
}

struct Board {
  let nails: Set<Point>
  let width: Int
  let height: Int

  static let slot: (Int) -> Int = { $0 / 2 + 1 }
  static let x: (Int) -> Int = { ($0 - 1) * 2 }

  init(_ nails: Set<Point>) {
    self.nails = nails
    self.width = nails.map { $0.x }.max()!
    self.height = nails.map { $0.y }.max()!
  }
}

struct Coin {
  let dirs: [Character]

  init(dirs: [Character]) {
    self.dirs = dirs
  }

  func simulate(_ board: Board, slot: Int) -> Int {
    var pos = Point(x: Board.x(slot), y: 0)

    var i = 0
    while pos.y <= board.height {
      if board.nails.contains(pos) {
        pos.x += {
          switch pos.x {
            case 0: return 1
            case board.width: return -1
            default: 
              return dirs[i] == "R" ? 1 : -1
          }
        }()
        i = (i + 1) % dirs.count
      }
      pos.y += 1
    }

    return Board.slot(pos.x)
  }
}

func getInput() -> (Board, [Coin]) {
  var grid: [[Character]] = []
  while let line = readLine(), !line.isEmpty {
    grid.append(Array(line))
  }

  var board: Set<Point> = []
  for (i, line) in grid.enumerated() {
    for (j, c) in line.enumerated() {
      if c == "*" { board.insert(Point(x: j, y: i)) }
    }
  }

  var coins: [Coin] = []
  while let line = readLine() {
    coins.append(Coin(dirs: Array(line)))
  }

  return (Board(board), coins)
}

let (board, coins) = getInput()

let ans = zip(1..., coins).map { 2 * $0.1.simulate(board, slot: $0.0) - $0.0 }.filter { $0 > 0 }.reduce(0, +)

print(ans)